addons/HotelPvP/scripts/Game/AFM_HotelPvP/GameMode/AFM_FactoryPvPGameMode.c

```diff
@@ -0,0 +1,442 @@
+		}
+		
+		if (defendersDead) 
+		{
+			GameEndAttackersWin();
+			return;
+		}
+	}
+	
+	protected bool AllHostagesFreed()
+	{
+		return GetGame().GetWorld().QueryEntitiesBySphere(
+				GetOrigin(),
+				m_iGameAreaRadius,
+				ProcessFoundHostages,
+				FilterHostageEntities,
+				EQueryEntitiesFlags.DYNAMIC
+		);
+	}
+	
+	protected bool ProcessFoundHostages(IEntity entity)
+	{
+		return false;
+	}
+	
+	protected bool FilterHostageEntities(IEntity entity)
+	{
+		if(entity.Type() != SCR_ChimeraCharacter)
+			return false; 
+		
+		FactionAffiliationComponent fac = FactionAffiliationComponent.Cast(entity.FindComponent(FactionAffiliationComponent));
+		return fac && fac.GetAffiliatedFactionKey() == "CIV";
+	}
+	
+	void TimeoutGameEnd()
+	{
+		if (!IsMaster())
+			return;
+		
+		PrintFormat("Timeout reached, ending game with {%1} faction win", m_TimeoutWinnerFactionKey);
+		SCR_Faction faction = SCR_Faction.Cast(m_FactionManager.GetFactionByKey(m_TimeoutWinnerFactionKey));
+		
+		
+		if (!faction)
+		{
+			GameEndTie();
+			return;
+		}
+	
+		GameEnd(m_TimeoutWinnerFactionKey);
+	}
+	
+	
+	protected void GameEnd(FactionKey fKey)
+	{
+		m_bHasGameStarted = false;
+		
+		Faction faction = m_FactionManager.GetFactionByKey(fKey);
+		int factionId = m_FactionManager.GetFactionIndex(faction);
+		SCR_GameModeEndData endData = SCR_GameModeEndData.CreateSimple(EGameOverTypes.ENDREASON_SCORELIMIT, winnerFactionId:factionId);
+		EndGameMode(endData);
+	}
+	
+	protected void GameEndTie()
+	{
+		Print("Game tie", LogLevel.NORMAL);
+		SCR_GameModeEndData endData = SCR_GameModeEndData.Create(
+			EGameOverTypes.FACTION_DRAW,
+			{}, {}
+		);
+		m_bHasGameStarted = false;
+		EndGameMode(endData);
+	}
+	
+	protected void GameEndDefendersWin()
+	{
+		Print("Defenders win!");
+		GameEnd(m_sDefenderFactionKey);
+	}
+	
+	protected void GameEndAttackersWin()
+	{
+		Print("Attackers win!");
+		GameEnd(m_sAttackerFactionKey);
+	}
+	
+	int GetFactionRemainingPlayers(FactionKey fKey)
+	{
+		SCR_Faction faction = SCR_Faction.Cast(m_FactionManager.GetFactionByKey(fKey));
+		if (!faction)
+		{
+			PrintFormat("Could not find faction %1", fKey, level:LogLevel.WARNING);
+			return 0;
+		}
+		array<int> playerIds = new array<int>;
+		
+		faction.GetPlayersInFaction(playerIds);
+		int remainingPlayers = 0;
+		
+		foreach(int id: playerIds)
+		{
+			PlayerController pc = GetGame().GetPlayerManager().GetPlayerController(id);
+			if (pc)
+			{
+				SCR_ChimeraCharacter ent = SCR_ChimeraCharacter.Cast(pc.GetControlledEntity());		
+				AFM_SpectatorComponent spectator = AFM_SpectatorComponent.Cast(pc.FindComponent(AFM_SpectatorComponent));
+				SCR_DamageManagerComponent damageManager = ent.GetDamageManager();
+				if (!damageManager.IsDestroyed() && !spectator.GetState()) {
+					remainingPlayers++;
+				}
+			}
+		}
+		PrintFormat("Found %1 players in faction %2", remainingPlayers, fKey, level: LogLevel.SPAM);
+		return remainingPlayers;
+	}
+	
+	void HandlePlayerDeath(IEntity unit)
```

Comment:
I feel like it would be convenient to know if methods are executed server or client side, might be obvious now but less so once you'll come back to the code after some break. I'm not sure if we should figure out some good method naming for that or simply have docs like this:
```suggestion
	//------------------------------------------------------------------------------------------------
	//! Server side handling of player deaths
	void HandlePlayerDeath(IEntity unit)
```
There's a plugin to generate doxygen headers for method which makes it a bit faster.

Reply:
I've added comments to some of the method and separated bits of gamemode component with `//------` separators, but I'm not really happy with this solution